GitTfs/Core/GitRepository.cs

```diff
@@ -520,18 +528,23 @@ public bool CreateBranch(string gitBranchName, string target)
             return reference != null;
         }
 
-        private readonly Dictionary<long, string> changesetsCache = new Dictionary<long, string>();
+        private readonly Dictionary<long, IList<string>> changesetsCache = new Dictionary<long, IList<string>>();
         private bool cacheIsFull = false;
 
         public string FindCommitHashByChangesetId(long changesetId, string tfsPath)
         {
-            var commit = FindCommitByChangesetId(changesetId, tfsPath);
+            var commit = FindCommitsByChangesetId(changesetId).FirstOrDefault(c => c.Message.Contains(tfsPath));
```

Comment:
>if the path is not empty, return immediatly if the commit with the good id and path is found
if the path is empty, continue to look for this specific changeset id through all the commits

We can not do this, because then `changesetsCache` will be corrupted, eg. some commits will never be found.